File: crates/codegraph-server/tests/fixtures/security/vulnerable_app.js
/**
 * Intentionally vulnerable Node.js/Express app for security scanner testing.
 * Every function triggers at least one CWE detection.
 */
const express = require('express');
const { exec } = require('child_process');
const mysql = require('mysql');
const fs = require('fs');
const crypto = require('crypto');

const app = express();

// CWE-798: Hardcoded credentials
const DB_PASSWORD = "password123";
const JWT_SECRET = "mysupersecretkey";

// === security_scan patterns ===

// CWE-78: Command injection
function runCommand(userInput) {
    exec(userInput, (err, stdout) => {
        console.log(stdout);
    });
}

// CWE-94: Code injection
function dynamicEval(code) {
    return eval(code);
}

// CWE-79: XSS
function renderHtml(userContent) {
    document.getElementById('output').innerHTML = userContent;
    document.write(userContent);
}

// CWE-89: SQL injection
function getUser(userId) {
    const conn = mysql.createConnection({ host: 'localhost' });
    conn.query("SELECT * FROM users WHERE id = " + userId);
}

// CWE-328: Weak hash
function hashPassword(password) {
    return crypto.createHash('md5').update(password).digest('hex');
}

// CWE-338: Weak PRNG
function generateToken() {
    return Math.random().toString(36).substring(2);
}

// === security_check_unchecked_returns (CWE-252) ===

function processData(data) {
    validateInput(data);
    sanitize(data);
    checkPermissions(data);
    const result = transform(data);
    return result;
}

function validateInput(data) { return data != null; }
function sanitize(data) { return data.trim(); }
function checkPermissions(data) { return true; }
function transform(data) { return data.toUpperCase(); }

// === security_check_resource_leaks (CWE-772) ===

function leakyFileRead(path) {
    const fd = fs.openSync(path, 'r');
    const buf = Buffer.alloc(4096);
    fs.readSync(fd, buf);
    // fd never closed
    return buf.toString();
}

function leakyConnection(host) {
    const net = require('net');
    const socket = net.connect({ host, port: 80 });
    socket.write('GET / HTTP/1.1\r\n\r\n');
    // socket never closed
}

// === security_check_misconfig (CWE-16) ===

app.use((req, res, next) => {
    // CWE-16: CORS wildcard
    res.header('Access-Control-Allow-Origin', '*');
    next();
});

app.get('/api/data', (req, res) => {
    // CWE-614/1004: Insecure cookie
    res.cookie('session', 'abc123', { secure: false, httpOnly: false });
    res.json({ data: 'secret' });
});

// SSL bypass
// const agent = new https.Agent({ rejectUnauthorized: false });

// === security_check_input_validation (CWE-20) ===

app.get('/api/users/:id', (req, res) => {
    const userId = req.params.id;
    // Direct use in SQL without validation
    getUser(userId);
    res.json({ ok: true });
});

app.get('/api/files', (req, res) => {
    // CWE-22: Path traversal
    const filename = req.query.name;
    const data = fs.readFileSync('/data/' + filename, 'utf8');
    res.send(data);
});

// === security_check_error_exposure (CWE-209) ===

app.get('/api/process', (req, res) => {
    try {
        const result = JSON.parse(req.body);
        res.json(result);
    } catch (e) {
        // Leaks stack trace to client
        res.status(500).json({
            error: e.message,
            stack: e.stack
        });
    }
});

app.get('/api/login', (req, res) => {
    try {
        const user = req.query.user;
        if (hashPassword(req.query.pass) !== DB_PASSWORD) {
            throw new Error('Auth failed for ' + user);
        }
    } catch (e) {
        res.status(401).send(e.toString());
    }
});

// === Taint flow ===

app.get('/api/exec', (req, res) => {
    const cmd = req.query.cmd;
    runCommand(cmd);
    res.send('done');
});

app.listen(3000);

File: crates/codegraph-server/tests/fixtures/security/vulnerable_app.py
"""
Intentionally vulnerable Python app for security scanner testing.
Every function triggers at least one CWE detection.
"""
import os
import pickle
import hashlib
import subprocess
import sqlite3
import yaml
import logging
from flask import Flask, request, jsonify, make_response

app = Flask(__name__)
DEBUG = True  # CWE-16: Security misconfiguration

# CWE-798: Hardcoded credentials
DB_PASSWORD = "super_secret_password_123"
API_KEY = "sk-1234567890abcdef1234567890abcdef"


# === security_scan patterns ===

def unsafe_query(user_input):
    """CWE-89: SQL Injection"""
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM users WHERE name = '" + user_input + "'")
    return cursor.fetchall()


def run_command(cmd):
    """CWE-78: OS Command Injection"""
    os.system(cmd)
    result = subprocess.Popen(cmd, shell=True)
    return result


def render_unsafe(user_html):
    """CWE-94: Code Injection"""
    eval(user_html)
    exec("print(" + user_html + ")")


def load_untrusted(data):
    """CWE-502: Deserialization"""
    obj = pickle.loads(data)
    config = yaml.load(data)
    return obj, config


def weak_hash(password):
    """CWE-328: Weak hash"""
    return hashlib.md5(password.encode()).hexdigest()


def weak_random():
    """CWE-338: Weak PRNG"""
    import random
    return random.random()


# === security_check_unchecked_returns (CWE-252) ===

def process_data(data):
    """Return values silently ignored"""
    validate_input(data)
    sanitize(data)
    check_permissions(data)
    result = transform(data)
    return result


def validate_input(data):
    return data is not None


def sanitize(data):
    return data.strip()


def check_permissions(data):
    return True


def transform(data):
    return data.upper()


# === security_check_resource_leaks (CWE-772) ===

def leaky_file_read(path):
    """Opens file without closing"""
    f = open(path, 'r')
    data = f.read()
    return data


def leaky_connection(host):
    """Opens socket without closing"""
    import socket
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((host, 80))
    s.send(b"GET / HTTP/1.1\r\n\r\n")
    return s.recv(4096)


def leaky_db(dsn):
    """Opens DB connection without closing"""
    conn = sqlite3.connect(dsn)
    cursor = conn.cursor()
    cursor.execute("SELECT 1")
    return cursor.fetchone()


# === security_check_misconfig (CWE-16) ===

@app.route("/api/data")
def cors_misconfigured():
    """CORS wildcard + insecure cookies"""
    response = make_response(jsonify({"data": "secret"}))
    response.headers["Access-Control-Allow-Origin"] = "*"
    response.set_cookie("session", "abc123", secure=False, httponly=False)
    return response


def ssl_bypass():
    """SSL verification disabled"""
    import requests
    return requests.get("http://internal-api.local/data", verify=False)


# === security_check_input_validation (CWE-20) ===

@app.route("/api/users/<user_id>")
def get_user(user_id):
    """Parameter used without validation"""
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM users WHERE id = " + user_id)
    return jsonify(cursor.fetchone())


@app.route("/api/files")
def read_file():
    """Path traversal — unvalidated file path"""
    filename = request.args.get("name")
    with open("/data/" + filename, "r") as f:
        return f.read()


# === security_check_error_exposure (CWE-209) ===

@app.route("/api/process")
def process_request():
    """Stack trace leaked to user"""
    try:
        data = request.get_json()
        result = unsafe_query(data["query"])
        return jsonify(result)
    except Exception as e:
        import traceback
        return jsonify({
            "error": str(e),
            "trace": traceback.format_exc(),
            "debug": repr(e)
        }), 500


@app.route("/api/login")
def login():
    """Internal error details in response"""
    try:
        user = request.args.get("user")
        password = request.args.get("pass")
        if weak_hash(password) != DB_PASSWORD:
            raise ValueError("Auth failed for " + user)
    except Exception as e:
        return str(e), 401


# === Taint flow: user input → dangerous sink ===

@app.route("/api/exec")
def execute():
    """Full taint chain: request → command execution"""
    cmd = request.args.get("cmd")
    run_command(cmd)
    return "done"


@app.route("/api/search")
def search():
    """Full taint chain: request → SQL query"""
    query = request.args.get("q")
    results = unsafe_query(query)
    return jsonify(results)


if __name__ == "__main__":
    app.run(debug=True, host="0.0.0.0")
